clients/drcachesim/tools/instr_decode_cache.h

```diff
@@ -0,0 +1,172 @@
+/* **********************************************************
+ * Copyright (c) 2024 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+ * instr_decode_cache.h: Library that supports caching of instruction decode
+ * information.
+ */
+
+#ifndef _INSTR_DECODE_CACHE_H_
+#define _INSTR_DECODE_CACHE_H_ 1
+
+#include "dr_api.h"
+#include "memref.h"
+
+#include <unordered_map>
+
+namespace dynamorio {
+namespace drmemtrace {
+
+/**
+ * Base class for storing instruction decode info. Users should sub-class this
+ * base class and implement set_decode_info() to derive and store the decode info
+ * they need.
+ *
+ * Derived classes must provide a default constructor to make objects
+ * corresponding to an invalid decoding.
```

Comment:
Have you looked at opcode_mix and ensured this interface works well there?

Reply:
opcode_mix just needs the opcode and category, so yes it should work there too. The only complexity with opcode mix is that it still tries to handle traces without encodings, so opcode_mix will still need to have decoding logic for that case.

Reply:
Added a TODO for now.

Reply:
If no-encoding traces are still a possibility (very much so w/ our public v1 out there): wouldn't you expect a decoding helper library to include the messy logic to handle those?  If I were a user I would not be happy with a library that only works on some traces when the code to handle all traces is sitting there in opcode_mix: I would think "why isn't that in this library".

If we aggressively deprecate (as in delete) public v1, and can make the claim no one should encounter an encoding-free trace, then not tackling that in this library seems reasonable.

Reply:
> If we aggressively deprecate (as in delete) public v1, and can make the claim no one should encounter an encoding-free trace, then not tackling that in this library seems reasonable.

But if we do that, we should remove the module mapping code from opcode_mix, right?  I guess for me it feels weird to have opcode_mix have that code yet this library not handle that.

Reply:
Supporting backward-compatibility is definitely useful. But that is not as clear when it comes at a non-trivial cost to the new interfaces and the user.

Including the module mapping logic in instr_decode_cache_t comes with more additions to the API that are not trivial, particularly: providing the file type (which isn't available at construction time, so the user would need to remember to provide that for instr_decode_cache_t to figure out what decode logic to use), also the module map init logic is not trivial either.

Since trace encodings have been around 2+ years, and we have the public v2 also coming up very soon, it doesn't seem unreasonable to me for new convenience libraries (decoding is still possible without instr_decode_cache_t) to make assumptions about availability of trace encodings.

However: I agree that it is important to let the user know in case they're using instr_decode_cache_t on an old trace. We could perhaps add some heuristic to detect that: many consecutive decodes all with zeroes?

Reply:
That said, if we're unable to resolve concerns about potential misuses of instr_decode_cache_t, it may be prudent to instead add it only to private libraries where we can make stronger assumptions about trace versions.

Reply:
> We could perhaps add some heuristic to detect that: many consecutive decodes all with zeroes?

I think we don't need a heuristic. Instead: since the reader sets encoding_is_new when it sees any encoding

https://github.com/DynamoRIO/dynamorio/blob/eec4721cf0b3856905f4565194b57751a76f90ea/clients/drcachesim/reader/reader.cpp#L231

maybe we can just use that: add to the decode cache only if encoding_is_new. That way, the cache would return nullptr for all instrs for traces without encodings, which is easy enough to spot.

Reply:
Let's make sure that field is always set: it looks like in general memref_t is uninitialized (nobody every fills it with zeroes: analyzer.cpp just declares the struct which has no constructor or init values; scheduler just copies what the reader gives it; reader only sets certain fields).

Reply:
If encoding_is_new may be a garbage value in some cases, isn't that an independent and bigger problem? Do we need a separate issue filed?

>  reader only sets certain fields

encoding_is_new is the only relevant field here though. I see the reader does set it to false: https://github.com/DynamoRIO/dynamorio/blob/73aae264aa7a41bf930b110572ea8f91add93d1c/clients/drcachesim/reader/reader.cpp#L234

> analyzer.cpp just declares the struct which has no constructor or init values

Can you point me to this code loc, where presumably the encoding_is_new is left unset?

Reply:
> If encoding_is_new may be a garbage value in some cases, isn't that an independent and bigger problem? Do we need a separate issue filed?

No, the docs say these fields are only valid when the filetype says there are encodings.  Looking at the docs: it explicitly says that for `.encoding`.  It seems implied for `.encoding_is_new` as it refers back to `.encoding`.

Reply:
It does seem like reader.cpp sets encoding_is_new every time: but it does not set .encoding which is uninitialized bits if there are no encodings available.  However, the docs don't guarantee that encoding_is_new is always set, so this would be relying on an implementation detail.

Reply:
We could change the docs to guarantee encoding_Is_new is set, if we want to rely on it; as is I don't think we can.

Reply:
Dropped the encoding_is_new idea and the new API checks filetype.